### PR TITLE
fix: use correct source maps in stacktrace

### LINKED ISSUE #2024: Error shown with line:column number that are off (source map not used)
### Describe the bug

Logged failing tests have the wrong line:column number.

### Reproduction

https://github.com/imba/imba-vite-template/tree/vitest-repro

Install dependencies and run npm test to see failing tests.

### System Info

```shell
System:
    OS: macOS 12.4
    CPU: (10) arm64 Apple M1 Max
    Memory: 197.63 MB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.14.2 - /usr/local/bin/node
    Yarn: 1.22.17 - /usr/local/bin/yarn
    npm: 8.5.0 - /usr/local/bin/npm
  Browsers:
    Chrome: 105.0.5195.102
    Chrome Canary: 107.0.5293.0
    Safari: 15.5
  npmPackages:
    vite: ^3.1.0 => 3.1.0 
    vitest: ^0.23.1 => 0.23.1
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by haikyuu**
use vitenode server to get a fetch result instead of a transform result.
And then get the sourcemap from it.

This fixes #2024

### CODE REVIEW

File: packages/vitest/src/utils/source-map.ts [hunk 1]
```diff
@@ -25,10 +25,14 @@ export async function interpretSourcePos(stackFrames: ParsedStack[], ctx: Vitest
   for (const frame of stackFrames) {
     if ('sourcePos' in frame)
       continue
-    const transformResult = ctx.server.moduleGraph.getModuleById(frame.file)?.ssrTransformResult
-    if (!transformResult)
+    const ssrTransformResult = ctx.server.moduleGraph.getModuleById(frame.file)?.ssrTransformResult
+    const transformResult = ctx.server.moduleGraph.getModuleById(frame.file)?.transformResult
+    const result = ssrTransformResult || transformResult
```
Comment (sheremet-va) [inline packages/vitest/src/utils/source-map.ts:30]: Why do we need this, if it's not used later? Maybe you wanted to put it on L#34?

**Review by sheremet-va — CHANGES_REQUESTED** [review]
Can you add a test for this, please?

File: packages/vitest/src/utils/source-map.ts (hunk 1 above)
Comment (haikyuu) [reply packages/vitest/src/utils/source-map.ts:30]: It's only used in the check `if (!result) continue`
The rationale behind it is that there is no need to calculate the position if the file hasn't been transformed.

And sometimes, the file has been transformed but `ssrTransformResult` is unedfined while `transformResult` isn't.

Comment (haikyuu) [reply packages/vitest/src/utils/source-map.ts:30]: The reason I didn't put it in line `#34` is that it has the wrong sourcemap. 

I think it's the sourcemap of transforming files to run in vite-node context where imports are mapped differently, and that's not the only one, so it has different sourcemaps.

**Comment by haikyuu** [comment]
> Can you add a test for this, please?

Sure, is there a test example for errors/stacktraces I can check? @sheremet-va

File: packages/vitest/src/utils/source-map.ts (hunk 1 above)
Comment (sheremet-va) [reply packages/vitest/src/utils/source-map.ts:30]: But it means `map` on L35 can be `undefined`? I don't really see a point in this check, tbh. It makes more sense to `continue` on L34, if map is not defined.

**Comment by sheremet-va** [comment]
> > Can you add a test for this, please?
> 
> Sure, is there a test example for errors/stacktraces I can check? @sheremet-va

To be honest, I don't know.

**Comment by haikyuu** [comment]
The test fails because of this error https://github.com/vitest-dev/vitest/runs/8301687675?check_suite_focus=true#step:7:668

```
Error: Failed to parse source for import analysis because the content contains invalid JS syntax. 
You may need to install appropriate plugins to handle the .imba file format, 
or if it's an asset, add \\"**/*.imba\\" to \`assetsInclude\` in your configuration.
```

I used my complete plugin to generate the snapshot, but I created a small version of the plugin just for the test. But I don't understand why Vite is complaining about this error. I'd use some help here 🆘

**Comment by sheremet-va** [comment]
> I used my complete plugin to generate the snapshot, but I created a small version of the plugin just for the test. But I don't understand why Vite is complaining about this error. I'd use some help here 🆘

Try adding `enforce: 'pre'`

**Comment by haikyuu** [comment]
Thanks @sheremet-va I was including extra quotes because I copied the value from the debugger. Fixed now :)

**Comment by haikyuu** [comment]
@sheremet-va The tests are failing because of another error  in https://github.com/vitest-dev/vitest/blob/main/test/fails/fixtures/timeout.test.ts#L4 that's not related to my PR

```
 Snapshots  1 failed
  - Expected   ""Error: Test timed out in 10ms.""
  + Received   ""Error: Hook timed out in 10ms.""
```

File: packages/vitest/src/utils/source-map.ts (hunk 1 above)
Comment (haikyuu) [reply packages/vitest/src/utils/source-map.ts]: Makes sense @sheremet-va Done 👍

File: test/stacktraces/test/runner.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      })
+        .catch((e) => {
+          error = e
+        })
+
+      expect(error).toBeTruthy()
+      const msg = String(error)
+        .split(/\n/g)
+        .reduce((acc, line) => {
+          if (line.includes('Start at') || line.includes('Duration') || line.includes('(1 test | 1 failed)') || line.includes(root))
```
Comment (sheremet-va) [inline test/stacktraces/test/runner.test.ts:33]: Maybe we should only include this part in snapshot?

Screenshot 2022-09-13 at 10 36 57

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: test/stacktraces/test/runner.test.ts (hunk 2 above)
Comment (haikyuu) [reply test/stacktraces/test/runner.test.ts:33]: Agree, will do

Comment (haikyuu) [reply test/stacktraces/test/runner.test.ts]: Done, they look nicer now

**Comment by haikyuu** [comment]
@sheremet-va now one test which is not related to my changes seems flaky: https://github.com/vitest-dev/vitest/blob/main/test/fails/test/runner.test.ts

```
  - Expected   ""Error: Test timed out in 10ms.""
  + Received   ""Error: Hook timed out in 10ms.""
```

**Comment by haikyuu** [comment]
@sheremet-va can you take a look again? I fixed the previous failing test. But I'm not sure why single threaded tests are failing in Ubuntu?

**Review by sheremet-va — APPROVED** [review]

**Pull request merged**